horace_core/sqw/@SQWDnDBase/private/smooth_dnd.m

```diff
@@ -0,0 +1,157 @@
+function dout = smooth_dnd(din, xunit, varargin)
+% Smooths a 1,2,3 or 4 dimensional dataset
+%
+%Syntax:
+%   >> dout = smooth_dnd (din, xunit, width, shape)
+%
+% Input:
+% ------
+%   din     Input dataset structure (data_sqw_dnd)
+%   xunit   If true, unit of measure is length of a unit along each axis
+%           If false, unit of measure is pixels
+%   width   Vector that sets the extent of the smoothing along each dimension.
+%          The interpretation of width depends on the argument 'shape' described
+%          below.
+%           If width is scalar, then the value is applied to all dimensions
+%
+%           e.g. if din is a 3-dimensional dataset, valid arguments for width might be:
+%                width = [2,4,5]    % 2, 3, 5 along the 1st, 2nd and 3rd dimensions
+%                width = 4.5        % 4.5 applied to all dimensions
+%           Invalid choices for 3-dimensions are
+%                width = [2,3]      % invalid number of dimensions
+%
+%   shape   Shape of smoothing function
+%               'hat'           Hat function
+%                                   - width gives FWHH along each dimension
+%                                   - width = 1,3,5,...;  n=0 or 1 => no smoothing
+%               'gaussian'      Gaussian; width gives FWHH along each dimension in pixels
+%                                   - elements where more than 2% of peak intensity
+%                                     are retained
+%
+%               'resolution'    Correlated Gaussian - 2D only (suitable for e.g. powder data)
+%
+% Output:
+% -------
+%   dout    Smoothed data structure
+
+
+% Original author: T.G.Perring
+%
+% $Revision:: 1759 ($Date:: 2020-02-10 16:06:00 +0000 (Mon, 10 Feb 2020) $)
+%
+% Horace v0.1   J. van Duijn, T.G.Perring
+
+
+% List available functions and set defaults. If more functions are to be
+% added as smoothing options, then place in the Horace private directory
+shapes = {'hat'; 'gaussian'; 'resolution'};       % internally available functions for convolution
+shape_handle = {@smooth_func_hat; @smooth_func_gaussian; @smooth_func_resolution};    % corresponding function handles
+width_default = 3;
+shape_default = 'hat';
+ishape_default = 1;
+
+
+% Check input parameters
+% ----------------------
+ndim = length(din.pax);   % no. dimensions of the data
+
+% Get width(s)
+if nargin>=3
+    width = varargin{1};
+elseif ~xunit
+    width = width_default*ones(1,ndim);
+else
+    error('HORACE:smooth:invalid_arguments', ...
+        'Must give smoothing parameter(s) if smoothing units')
```

Comment:
I think this is implying that if xunit is not present it will default to false. But it does not look right. The final else appears to say that if I call `smooth_dnd(din) then I will have xunit==false and default widths of default*ones(), otherwise I must put in widths for my smoothing in units. It would be nicer to make xunit compulsory. As the only calls currently are in smooth and smooth_units, both of which have all 3 args, this would be good to do now.